executor/partition_table_test.go

```diff
@@ -673,3 +673,101 @@ func (s *globalIndexSuite) TestIssue21731(c *C) {
 }
+
+func (s *testSuiteWithData) TestRangePartitionBoundariesLtM(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+
+	tk.MustExec("set @@tidb_partition_prune_mode = 'dynamic'")
+	tk.MustExec("create database TestRangePartitionBoundariesLtM")
+	defer tk.MustExec("drop database TestRangePartitionBoundariesLtM")
+	tk.MustExec("use TestRangePartitionBoundariesLtM")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec(`CREATE TABLE t
+(a INT, b varchar(255))
+PARTITION BY RANGE (a) (
+ PARTITION p0 VALUES LESS THAN (1000000),
+ PARTITION p1 VALUES LESS THAN (2000000),
+ PARTITION p2 VALUES LESS THAN (3000000))`)
+
+	var input []string
+	var output []struct {
+		SQL  string
+		Plan []string
+		Res  []string
+	}
+	s.testData.GetTestCases(c, &input, &output)
+	for i, tt := range input {
+		var isSelect bool = false
+		if strings.HasPrefix(strings.ToLower(tt), "select ") {
+			isSelect = true
+		}
+		s.testData.OnRecord(func() {
+			output[i].SQL = tt
+			if isSelect {
+				output[i].Plan = s.testData.ConvertRowsToStrings(tk.MustQuery("explain format = 'brief' " + tt).Rows())
+				output[i].Res = s.testData.ConvertRowsToStrings(tk.MustQuery(tt).Sort().Rows())
+			} else {
+				output[i].Plan = nil
+				output[i].Res = nil
+			}
+		})
+		if isSelect {
+			tk.MustQuery("explain format = 'brief' " + tt).Check(testkit.Rows(output[i].Plan...))
+			tk.MustQuery(tt).Sort().Check(testkit.Rows(output[i].Res...))
+		} else {
+			tk.MustExec(tt)
+			c.Assert(output[i].Res, IsNil)
+		}
+	}
+}
+
+func (s *testSuiteWithData) TestRangePartitionBoundariesLtS(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+
+	tk.MustExec("set @@tidb_partition_prune_mode = 'dynamic'")
+	tk.MustExec("create database TestRangePartitionBoundariesLtS")
+	defer tk.MustExec("drop database TestRangePartitionBoundariesLtS")
+	tk.MustExec("use TestRangePartitionBoundariesLtS")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec(`CREATE TABLE t
+(a INT, b varchar(255))
+PARTITION BY RANGE (a) (
+ PARTITION p0 VALUES LESS THAN (1),
+ PARTITION p1 VALUES LESS THAN (2),
+ PARTITION p2 VALUES LESS THAN (3),
+ PARTITION p3 VALUES LESS THAN (4),
+ PARTITION p4 VALUES LESS THAN (5),
+ PARTITION p5 VALUES LESS THAN (6),
+ PARTITION p6 VALUES LESS THAN (7))`)
+
+	var input []string
+	var output []struct {
+		SQL  string
+		Plan []string
+		Res  []string
+	}
+	s.testData.GetTestCases(c, &input, &output)
+	for i, tt := range input {
+		var isSelect bool = false
+		if strings.HasPrefix(strings.ToLower(tt), "select ") {
+			isSelect = true
+		}
+		s.testData.OnRecord(func() {
+			output[i].SQL = tt
+			if isSelect {
+				output[i].Plan = s.testData.ConvertRowsToStrings(tk.MustQuery("explain format = 'brief' " + tt).Rows())
+				output[i].Res = s.testData.ConvertRowsToStrings(tk.MustQuery(tt).Sort().Rows())
+			} else {
+				output[i].Plan = nil
+				output[i].Res = nil
+			}
+		})
+		if isSelect {
+			tk.MustQuery("explain format = 'brief' " + tt).Check(testkit.Rows(output[i].Plan...))
```

Comment:
How about using `tk.MustPartition` to just check the pruned partition instead of the whole plan? @mjonss

Reply:
I think that would be good for white box testing (i.e. verifying the implementation, with full knowledge about specific implementation), but I have not yet full understanding, and having several tests where one manually needs to update the code for the result takes a lot of time (Now I can only run `(cd executor ; go test -check.f TestRangePartitionBound --record)` to update the result file (executor/testdata/executor_suite_out.json file) and verify the correct partitioning pruning.

I still try to understand the preferred way to implement tests for different scenarios and are happy to discuss it further and have clear guidelines on when to write/implement which tests, but I think it is outside this PR.
Should I rework the PR and use tk.MustPartition for each test statement?

Reply:
Yes, I think the goal of these test cases is to check whether partition pruning is working and can return the correct result under these situations, so I think using `MustPartition` is clearer and enough.
And the whole plan contains some info unrelated to partition pruning, for example, the `est-row`, so if someone updates the behavior of our estimation module, these cases will also be affected and he has to update them as well, which is a little inconvenient to maintain.

Reply:
What if I add a function to `util/testkit/testkit.go` that instead of the full explain, just outputs the used partitions instead? Or rather {'all'|'dual'|<list of partitions used>} in the explain query (and for now, leave the handling of multiple different pruning sets, due to JOIN/UNION of multiple partitioned tables, for later)?
Then it would only check for the pruning, while still be able to separate the return to the _out.json file for easier management of the test and results.

Reply:
OKay, that sounds reasonable for me. @mjonss

Reply:
@mjonss Friendly ping, please address this comment. Hope we can merge it within 2 days~

Reply:
This has been done, see `UsedPartitions()` in `util/testkit/testkit.go`, so I consider this resolved.